### PR TITLE
Clear Last Data from Signal

### PR DESCRIPTION
**Pull request by netizen01**
lastDataFired is a strong reference to the last object that was sent through the Signal. If sending objects with a wide reference graph, this can lead to some unexpected memory "leaks".

Ideally the lastDataFired should be a weak reference, but that is not possible with Swift non-class types.

This can already be achieved by setting fireCount to 0 and the lastDataFired to nil, but an explicit API call to do some clean up is probably a better design.

### CODE REVIEW

**Comment by artman** [comment]
Thanks for the contribution. This is indeed a shortcoming and potential problem, and I think the correct way to fix this is to make storing the previously fired data opt-in, as the normally you wouldn't want to keep it around.

This would be a breaking change and bump the library versions to 3.0.

**Comment by netizen01** [comment]
Agreed. The complete PR here does have the previous fired data as an opt-in (see d7b5bc5).

Love this framework. Use it everywhere. Thanks!

File: Signals/Signal.swift [hunk 1]
```diff
@@ -18,6 +18,9 @@ public class Signal<T> {
     /// The last data that the signal was fired with.
     public var lastDataFired: T? = nil
     
+    /// Should the Signal retain a reference to the last data that it was fired with?
```
Comment (artman) [inline Signals/Signal.swift]: You should refrain from using questions as documentation. A better wording here would be:

`Whether or not the Signal should retain a reference to the last data it was fired with. Defaults to false.`

File: Signals/Signal.swift [hunk 2]
```diff
@@ -30,8 +33,12 @@ public class Signal<T> {
         }
     }
     
-    public init() {
+    /// Signal Initializer
+    /// 
+    /// - parameter retainLastData: Should the Signal retain a reference to the last data that it was fired with?
+    public init(retainLastData: Bool = false) {
```
Comment (artman) [inline Signals/Signal.swift]: Same here: `- parameter retainLastDatA: Whether or not the Signal should retain a reference to the last data it was fired with. Defaults to false.`

File: Signals/Signal.swift [hunk 3]
```diff
@@ -78,8 +85,8 @@ public class Signal<T> {
     /// - parameter callback: The closure to invoke whenever the signal fires.
     public func listenPast(listener: AnyObject, callback: (T) -> Void) -> SignalListener<T> {
         let signalListener = self.listen(listener, callback: callback)
-        if fireCount > 0 {
-            signalListener.callback(lastDataFired!)
+        if let lastDataFired = lastDataFired where fireCount > 0 {
```
Comment (artman) [inline Signals/Signal.swift]: It's not really necessary to check fireCount here. If you have data you fire the signal.

File: Signals/Signal.swift [hunk 4]
```diff
@@ -92,8 +99,8 @@ public class Signal<T> {
     /// - parameter callback: The closure to invoke whenever the signal fires.
     public func listenPastOnce(listener: AnyObject, callback: (T) -> Void) -> SignalListener<T> {
         let signalListener = self.listen(listener, callback: callback)
-        if fireCount > 0 {
-            signalListener.callback(lastDataFired!)
+        if let lastDataFired = lastDataFired where fireCount > 0 {
```
Comment (artman) [inline Signals/Signal.swift]: Same here.

File: Signals/Signal.swift [hunk 5]
```diff
@@ -132,6 +139,12 @@ public class Signal<T> {
     public func removeAllListeners() {
         signalListeners.removeAll(keepCapacity: false)
     }
+    
+    /// Clears the last fired data from the Signal and resets the fire count
+    public func clearLastData() {
+        fireCount = 0
```
Comment (artman) [inline Signals/Signal.swift]: It's not clear that fireCount is set to 0 as a side-effect. I'd just leave it be.

File: SignalsTests/SignalsTests.swift [hunk 6]
```diff
@@ -73,7 +73,7 @@ class SignalsTests: XCTestCase {
         var lastArgument = 0
         
         emitter.onInt.listen(self, callback: { (argument) in
-            dispatchCount++
```
Comment (artman) [inline SignalsTests/SignalsTests.swift]: Fixes for these have already landed, you'd need to rebase on master in order to get out of the merge conflicts.

Comment (artman) [inline SignalsTests/SignalsTests.swift]: Fixes for these have already landed, you'd need to rebase on master in order to get out of the merge conflicts.

**Comment by artman** [comment]
Looking good! Have a few comments and the you need to rebase on master after which this is good to land.

**Comment by artman** [comment]
Implements #30 #

File: SignalsTests/SignalEmitter.swift [hunk 7]
```diff
@@ -10,8 +10,8 @@ import Foundation
...
 class SignalEmitter {
-    let onNoParams = Signal<()>()
-    let onInt = Signal<Int>()
-    let onString = Signal<String>()
-    let onIntAndString = Signal<(intArgument: Int, stringArgument: String)>()
+    let onNoParams = Signal<()>(retainLastData: true)
+    let onInt = Signal<Int>(retainLastData: true)
+    let onString = Signal<String>(retainLastData: true)
+    let onIntAndString = Signal<(intArgument: Int, stringArgument: String)>(retainLastData: true)
 }
```
Comment (artman) [inline SignalsTests/SignalEmitter.swift]: Nit: Newline is missing

File: Signals/Signal.swift [hunk 8]
```diff
@@ -18,6 +18,9 @@ public class Signal<T> {
     /// The last data that the signal was fired with.
     public var lastDataFired: T? = nil
     
+    /// Whether or not the Signal should retain a reference to the last data it was fired with. Defaults to false.
+    public var retainLastData: Bool = false
```
Comment (artman) [inline Signals/Signal.swift]: This actually should have a didSet hook that would clear out the retainedData if the value would be set to false.

**Comment by artman** [comment]
One last request: Two new unit test should be added to ensure that

1) When retainLastData is false, the previous data isn't retained
2) When retainLastData is set to false the previously fired data is cleared out.

File: SignalsTests/SignalEmitter.swift (hunk 7 above)
Comment (netizen01) [inline SignalsTests/SignalEmitter.swift]: ?

Comment (artman) [inline SignalsTests/SignalEmitter.swift]: Files should end with an empty line.

**Comment by artman** [comment]
Cheers, looks good. Lemme pile my changes on top of this and release 3.0.

**Pull request merged**